### PR TITLE
Fix S6 block, add ResidualBlock and improve S6

### PR DESCRIPTION
**Pull request by FilippoOlivo**
This PR align the initialization of dt to the one in the original GitHub repo of Mamba. In preliminary tests, this initialization (and making dt trainable) allow training convergence with S6. However, the test were done on very short sequences

### CODE REVIEW

File: ssm/model/block/residual_block.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import torch
+
+
+class ResidualBlock(torch.nn.Module):
```
Comment (GiovanniCanali) [inline ssm/model/block/residual_block.py:4]: Isn't it better to write something like `x + prediction` in the forward of the interested models, instead of creating a new class?

File: ssm/model/block/s6_block.py [hunk 2]
```diff
@@ -11,22 +12,26 @@ class DeltaNetwork(torch.nn.Module):
...
-        # Initialize the parameters
-        self.dt = dt
         self.input_dim = input_dim
-
-        # Define the layers
+        self.linear = torch.nn.Linear(input_dim, 1, bias=False)
         self.activation = torch.nn.Softplus()
-        self.linear = torch.nn.Linear(input_dim, 1)
+
+        dt = torch.exp(
```
Comment (GiovanniCanali) [inline ssm/model/block/s6_block.py:29]: Is there any reference for this change?

File: ssm/model/block/s6_block.py [hunk 3]
```diff
@@ -36,11 +41,13 @@ def forward(self, x):
...
-        delta = self.activation(self.linear(x)) + self.dt
-        return delta.repeat(1, 1, self.input_dim)
+
+        x = self.linear(x)
+        x = x.expand(-1, -1, self.input_dim)
+        return self.activation(x + self.bias)
 
 
-class S6Block(torch.nn.Module):
+class S6Block(torch.torch.nn.Module):
```
Comment (GiovanniCanali) [inline ssm/model/block/s6_block.py:50]: Duplicated `torch`

File: ssm/model/block/s6_block.py [hunk 4]
```diff
@@ -49,7 +56,7 @@ class S6Block(torch.nn.Module):
     parts of the input sequence, making it suitable for tasks such as selective
     copy.
 
-    The output is computed in an efficient manner by leveraging the parallel
+    The output is computed in an efficient matorch.nner by leveraging the parallel
```
Comment (GiovanniCanali) [inline ssm/model/block/s6_block.py:59]: Please, restore `manner`

File: ssm/model/block/s6_block.py [hunk 5]
```diff
@@ -87,16 +99,20 @@ def __init__(
         # Initialize parameters
         self.input_dim = input_dim
         self.hid_dim = hid_dim
-        self.dt = dt
 
         # Initialize the matrix A
         A = compute_S4DReal(hid_dim, real_random=real_random).unsqueeze(0)
-        self.A = torch.nn.Parameter(A.repeat(input_dim, 1))
+        self.A = torch.torch.nn.Parameter(A.repeat(input_dim, 1))
```
Comment (GiovanniCanali) [inline ssm/model/block/s6_block.py:105]: Duplicated `torch`

File: ssm/model/block/s6_block.py [hunk 6]
```diff
@@ -177,7 +193,7 @@ def _complex_log(input_):
         a_star = torch.cumsum(log_a, dim=1)
 
         # Padding over the sequence length L
-        a_star = torch.nn.functional.pad(a_star, (0, 0, 0, 0, 1, 0))
+        a_star = torch.torch.nn.functional.pad(a_star, (0, 0, 0, 0, 1, 0))
```
Comment (GiovanniCanali) [inline ssm/model/block/s6_block.py:196]: Duplicated `torch`

**Review by GiovanniCanali — CHANGES_REQUESTED** [review]
Hi @FilippoOlivo, thank you for the PR. 
I believe that residual block can be added to the models in a much simpler way: just add residual connections in the `forward` of interested models / blocks.
Be careful when using automatic substitutions.

File: ssm/model/block/residual_block.py (hunk 1 above)
Comment (FilippoOlivo) [reply ssm/model/block/residual_block.py:4]: I believe this solution is better due the fact that we use the residual on a sequence of modules 
SSM + mixing + activation (+ Layer Normalization). In this way adding the Residual is just a wrapping of a torch sequential. In the other case we need to break down the torch Sequential module in a torch.nn.Module list and call and iterate over the list of modules (and then adding the residual)

**Comment by FilippoOlivo** [comment]
@GiovanniCanali Thanks for the suggestion! I didn't notice the duplications, I will fix them soon. They I suggest to apply the same logic to S4. What do you think?

File: ssm/model/block/residual_block.py (hunk 1 above)
Comment (FilippoOlivo) [reply ssm/model/block/residual_block.py:4]: Maybe another solution could be creating a Block which handles residual, activation and LayerNormalization instead of only handling residual. This makes the code flexible without adding an "trivial" Module. @GiovanniCanali what do you think?

**Comment by GiovanniCanali** [comment]
> @GiovanniCanali Thanks for the suggestion! I didn't notice the duplications, I will fix them soon. They I suggest to apply the same logic to S4. What do you think?

If you refer to the handling of residual connections, it is ok for me. In that case, please use this very same PR for changes.
If you meant something else, please clarify.

File: ssm/model/block/s6_block.py (hunk 2 above)
Comment (FilippoOlivo) [reply ssm/model/block/s6_block.py:29]: I read more carefully mamba paper and, at page 9 they say 
_In our experiments, the Δ parameter (which can be viewed as a bias term) is initialized to 𝜏−1
Δ (Uniform([0.001,0.1])), following prior work on SSMs (Gu, Johnson, Timalsina, et al. 2023)._
Do we want to apply this precise strategy?

**Comment by FilippoOlivo** [comment]
@GiovanniCanali I modify the initialization of dt for all the S4 and S6 blocks and I add residual connection in S4. Ready to merge from my side

**Review by GiovanniCanali — APPROVED** [review]
Good to go!
